### PR TITLE
add bias ordering transform to improve identifiability in neural nets

### PR DESCRIPTION
**Pull request by mdmould**
Adds a reparameterization of a vector so that its elements are in increasing order. This is useful to apply to bias vectors in neural networks to remove degeneracy/multimodality in their likelihoods/posteriors induced by the label switching problem. See http://bayesiandeeplearning.org/2017/papers/15.pdf and https://num.pyro.ai/en/latest/_modules/numpyro/distributions/transforms.html#OrderedTransform.

I wasn't sure whether to add this to flowjax or paramax, as it looks you're planning to move the reparam functionality from flowjax to here.

### CODE REVIEW

**Comment by mdmould** [comment]
I realise now that this functionality can probably be achieved with the following:
```python
import jax
import jax.numpy as jnp
import equinox
from paramax.wrappers import Parameterize

def ordered_vector(vector):
    first, log_diffs = vector[..., :1], vector[..., 1:]
    first_and_diffs = jnp.concatenate([first, jnp.exp(log_diffs)], axis=-1)
    return jnp.cumsum(first_and_diffs, axis = -1)

linear = equinox.nn.Linear(
    in_features = 10,
    out_features = 10,
    use_bias = True,
    key = jax.random.key(0),
)

bias = Parameterize(ordered_vector, linear.bias)
linear = equinox.tree_at(lambda linear: linear.bias, linear, replace = bias)
```
However, when I try to use this in a normalizing flow example by doing tree surgery to replace all the biases in an already-instantiated flow, the Parameterized biases do not get unwrapped when the flow methods. I tried following the example of `WeightNormalization` in `block_autoregressive_linear` but I'm not sure what I'm missing. @danielward27 do you have any pointers of how to do this correctly?

**Comment by mdmould** [comment]
Never mind, it's working now - sorry for the spam!

**Comment by danielward27** [comment]
Nice, glad to hear you got it working. I think this pattern of using a function and passing it to ``Parameterize`` is generally better, unless there is need for an init method. A module of commonly used functions for parameterizations including ordered vector would be good, although I've not thought much about it yet. Ideally for common parameterizations, we should also have an "inverse" so that both  ``Parameterize(f, unbounded_x)`` and ``Parameterize(f, inv_f(constrained_x))`` are convenient.

**Comment by mdmould** [comment]
I think the Bijection class infrastructure provides this, though without the requirement of the log Jacobian determinant (though for this example that can be computed). Perhaps the function input to Parameterize could optionally be a Bijection instance so both forwards and inverse methods are available, without requiring the `*and_log_det` methods. It would then be up to the user to make them nontrainable if required. If this reparameterization functionality is going to be separated from flowjax it might be better for such (potentially non-differentiable) bijections to have their own abstract base to prevent circular dependencies.

**Comment by danielward27** [comment]
I used a similar approach initially in FlowJAX (based on bijections), but right now we can do ``Parameterize(bijection.transform, bijection.inverse(x))`` (with or without the inverse), which is quite clear, without introducing any dependency  (some consideration of vectorization is needed here, but that's another discussion).

I agree that a new class is better (or even just a named tuple), but it shouldn't be called bijection because all we (possibly) need is a right inverse e.g. ``softmax(jnp.log(probs_summing_to_one)) == probs_summing_to_one``.
